src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseInviteService.cs

```diff
@@ -0,0 +1,227 @@
+            Guid userId, string email, string publicationTitle)
+        {
+            // check the user doesn't already have the user release roles
+            var existingReleaseRoleReleaseIds = _contentDbContext.UserReleaseRoles
+                .AsQueryable()
+                .Where(urr =>
+                    releaseIds.Contains(urr.ReleaseId)
+                    && urr.Role == Contributor
+                    && urr.UserId == userId)
+                .Select(urr => urr.ReleaseId)
+                .ToList();
+
+            var missingReleaseRoleReleaseIds = releaseIds
+                .Where(releaseId => !existingReleaseRoleReleaseIds.Contains(releaseId))
+                .ToList();
+
+            if (missingReleaseRoleReleaseIds.IsNullOrEmpty())
+            {
+                return ValidationActionResult(UserAlreadyHasReleaseRoles);
+            }
+
+            var emailResult = await SendContributorInviteEmail(
+                publicationTitle: publicationTitle,
+                releaseIds: missingReleaseRoleReleaseIds,
+                email: email);
+            if (emailResult.IsLeft)
+            {
+                return emailResult;
+            }
+
+            await _userReleaseRoleRepository.CreateManyIfNotExists(
+                userId: userId,
+                releaseIds: missingReleaseRoleReleaseIds,
+                role: Contributor,
+                createdById: _userService.GetUserId()
+            );
+
+            await _userReleaseInviteRepository.CreateManyIfNotExists(
+                releaseIds: missingReleaseRoleReleaseIds,
+                email: email,
+                releaseRole: Contributor,
+                emailSent: true,
+                createdById: _userService.GetUserId(),
+                accepted: true);
+
+            return Unit.Instance;
+        }
+
+        private async Task<Either<ActionResult, Publication>> ValidateReleaseIds(Publication publication,
+            List<Guid> releaseIds)
+        {
+            var distinctReleaseIds = releaseIds.Distinct().ToList();
+            if (distinctReleaseIds.Count != releaseIds.Count)
+            {
+                throw new ArgumentException($"{nameof(releaseIds)} should not contain duplicates",
+                    nameof(releaseIds));
+            }
+
+            await _contentDbContext
+                .Entry(publication)
+                .Collection(p => p.Releases)
+                .LoadAsync();
+
+            var publicationReleaseIds = publication
+                .ListActiveReleases()
+                .Select(r => r.Id)
+                .ToList();
+
+            if (!releaseIds.All(publicationReleaseIds.Contains))
+            {
+                return ValidationActionResult(NotAllReleasesBelongToPublication);
+            }
+
+            return publication;
+        }
+
+        private async Task<Either<ActionResult, Unit>> SendContributorInviteEmail(
+            string publicationTitle, List<Guid> releaseIds, string email)
```

Comment:
Seems like we construct a bunch of emails already in `EmailTemplateService`, should this be in there?

Reply:
I didn't know `EmailTemplateService` was a thing. I was following what `PreReleaseUserService#SendPreReleaseInviteEmail` does.

In the instance of PreRelease, `SendPreReleaseInviteEmail` uses `PreReleaseService`, and if it was moved to `EmailTemplateService` either:
- that call to `_preReleaseService` would have to be done in `PreReleaseUserService` and passed to `_emailTemplateService.SendPreReleaseInviteEmail` as an argument.
- we'd need to add `PreReleaseService` to `EmailTemplateService`.

This is similarly true of `SendContributorInviteEmail`, as it requires release titles for all releaseIds fetching from the ContentDb.

Given that, I'm inclined to think that `Send*Email` methods can live in the services that use them, if they're unlikely to be called from more than one service. I checked the `Send*Email` methods in `EmailTemplateService`, and they're each all only called from one service. So my inclination is to have each service responsible for sending its own emails rather than having an `EmailTemplateService`.

All that said, I'm not sure I consider this an issue worth the time of changing things - it's probably not even worth the time I've spent writing this out. So I'd personally just leave things as they are. But happy to change things if others feel strongly about it.